File: ros2_foxy_tutorial_ws/src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
# ros client library: rcl
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64
 
class NumPubNode(Node): 
    def __init__(self):
        super().__init__("number_publisher") 

        # test functionality of declare_parameter
        #self.declare_parameter("test_param")
        #self.declare_parameter("second_test_param")

        self.declare_parameter("number_to_publish", 2)
        self.pub_number = self.get_parameter("number_to_publish").value
        # $ ros2 run my_py_pkg number_publisher --ros-args -p number_to_publish:=4
        
        self.declare_parameter("publish_frq", 2.0)
        self.publish_frq_ = self.get_parameter("publish_frq").value

        # self.pub_number = 2
        self.publisher_ = self.create_publisher(Int64, "number", 10)
        self.counter_ = self.create_timer(1.0/self.publish_frq_, self.publish_numbers)
        self.get_logger().info("Number publisher started")
 
    def publish_numbers(self):
        msg = Int64()
        
        msg.data = self.pub_number
        self.publisher_.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = NumPubNode() 
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: ros2_foxy_tutorial_ws/src/my_cpp_pkg_/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"

class NumberPublisherNode : public rclcpp::Node
{
public:
    NumberPublisherNode() : Node("number_publisher"), number_(2)
    {
        this->declare_parameter("name");

        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number", 10);
        timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                         std::bind(&NumberPublisherNode::publishNumber, this));
        RCLCPP_INFO(this->get_logger(), "Number publisher has been started.");
    }

private:
    void publishNumber()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = number_;
        publisher_->publish(msg);
    }

    int number_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}